ciao-image/datastore/ceph.go

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2016 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datastore
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path"
+
+	"github.com/01org/ciao/ciao-storage"
+)
+
+// Ceph implements the DataStore interface for Ceph RBD based storage
+type Ceph struct {
+	ImageTempDir string
+	BlockDriver  storage.CephDriver
+}
+
+// Write copies an image into the posix filesystem.
+// If the image already exists it will be overridden.
+func (c *Ceph) Write(ID string, body io.Reader) (int64, error) {
+	imagePath := path.Join(c.ImageTempDir, ID)
+
+	image, err := os.Create(imagePath)
+	if err != nil {
+		return 0, fmt.Errorf("Error creating temporary image file: %v", err)
+	}
+
+	// TODO(rbradford): Is there a better way
+	buf := make([]byte, 1<<16)
+	res, err := io.CopyBuffer(image, body, buf)
+
+	if err != nil {
+		return 0, fmt.Errorf("Error writing temporary image file: %v", err)
+	}
+
+	_, err = c.BlockDriver.CreateBlockDevice(ID, imagePath, 0)
+	if err != nil {
+		return 0, fmt.Errorf("Error creating block device: %v", err)
+	}
+
+	err = c.BlockDriver.CreateBlockDeviceSnapshot(ID, "ciao-image")
+	if err != nil {
+		c.BlockDriver.DeleteBlockDevice(ID)
+		return 0, fmt.Errorf("Unable to create snapshot: %v", err)
```

Comment:
Don't we leak imagePath here?  Maybe we should move the call to os.Remove up a little.